### DISCUSSION TITLE
Time scattering is sensitive to phase shift

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#731] [Q&A] [answered]
This should be the last thing it's sensitive to, yet it's about as sensitive to it as to a significant time shift. Further, larger `J` yields _less_ phase-shift invariance, especially in ratio to time-shift.

Effect diminishes with larger `f`, but is still too large for reasonable `f`.

```
(N, 2**J) = (8192, 4096)
L1L2(time -shifted):    0.09560047009553223
L1L2(phase-shifted):    0.05806556715006837
l1l2_phase / l1l2_time: 0.6073774228520451

(N, 2**J) = (8192, 2048)
L1L2(time -shifted):    0.3418317254157979
L1L2(phase-shifted):    0.033166546595322324
l1l2_phase / l1l2_time: 0.09702594618735033
```

![image](https://user-images.githubusercontent.com/16495490/115418109-440ce300-a20a-11eb-9fca-534938398f36.png)

<details>
  <summary><b>code</b></summary>

```python
import numpy as np
import scipy.signal
from kymatio.numpy import Scattering1D
from ssqueezepy import cwt
from ssqueezepy.visuals import imshow, plot

### helper methods 
def _l1l2(x):
    return np.sum(np.sqrt(np.mean(x**2, axis=1)), axis=0)

def l1l2_mod(x0, x1):
    """L1L2 agnostic to choice of reference."""
    return _l1l2(x1 - x0) / ( (_l1l2(x0) + _l1l2(x1))/2 )

#%% define signals
N = 8192
f = N // 32
seg_len = N // 6
phi = -np.pi / 2
shift = N // 6

t = np.linspace(0, 1, N, endpoint=False)
window = scipy.signal.tukey(seg_len, alpha=0.5)
pad_right = (N - len(window)) // 2
pad_left  = N - len(window) - pad_right
window = np.pad(window, [pad_left, pad_right])

x0  = window * np.cos(2*np.pi * f * t)
xps = window * np.cos(2*np.pi * f * t + phi)
xts = np.roll(x0, shift)

#%% visualize
kw = dict(abs=1, ticks=0)
imshow(cwt(x0)[0],  **kw, title="abs(CWT) | x")
imshow(cwt(xts)[0], **kw, title="abs(CWT) | x, time shifted")

zoom = int(.4 * N)
plot(x0[zoom:-zoom], ticks=0)
plot(xps[zoom:-zoom], ticks=0,
     title="x & x phase-shifted (90 deg), superimposed | zoomed", show=1)

#%% create scattering object
J = int(np.log2(N) - 1)
Q = 16
ts = Scattering1D(J=J, Q=Q, shape=N)

#%% scatter
ts_x0  = ts(x0)
ts_xps = ts(xps)
ts_xts = ts(xts)

#%% compare
l1l2_x0xps = l1l2_mod(ts_x0, ts_xps)
l1l2_x0xts = l1l2_mod(ts_x0, ts_xts)

print(("L1L2(time -shifted):    {}\n"
       "L1L2(phase-shifted):    {}\n"
       "l1l2_phase / l1l2_time: {}"
       ).format(l1l2_x0xts, l1l2_x0xps, l1l2_x0xps / l1l2_x0xts))
```
</details>

**Comment by lostanlen** [comment]
So, i've just re-read the relevant sections of the group invariant scattering paper and it seems like those mixed L1/L2 norms are only necessary for the unaveraged version of the scattering transform ("U").
For S, we can use simple L2 norms. (see definition before Equation 2.25)

```python
>>> np.linalg.norm(ts_x0[:, :]-ts_xps[:, :])/np.linalg.norm(ts_x0[:,:])
0.016878240727518007

```

Removing the zeroth-order makes it even better
```python
>>> np.linalg.norm(ts_x0[1:, :]-ts_xps[1:, :])/np.linalg.norm(ts_x0[1:,:])
0.009073002258991388

```

**Reply by lostanlen** [reply]
Compare with the L2 distance in the raw signal domain:

```python
>>> np.linalg.norm(x0 - xps)/np.linalg.norm(x0)
1.4142155188964185
```

(aka square root of 2)

**Reply by OverLordGoldDragon** [reply]
This made a large difference on FDTS (`f0, total_shift, seg_len, n_partials = N//12, N//12, N//8, 5`):

```
JTFS, TS, ratio
# N = 2048
0.148439 -- 0.073504 -- 2.019  # L1L2
0.231515 -- 0.039765 -- 5.822  # L2
# N = 8192
0.194839 -- 0.016496 -- 11.811  # L1L2
0.293174 -- 0.007113 -- 41.219  # L2

```

For phase vs timeshift, the ratios halve, now `0.375`, or `0.21` w/o zeroth-order. This is still too high, compare blue and orange (`xps`), it's equiv to a timeshift of `8`, compared to `1365` (`xts`)...

Also raw signal domain comparison gives nonsensical results, `norm(x0, xts) == norm(x0, xps)`? (this _is_ accurate per the metric)

> i've re-read the relevant parts of

Page #s?

**Reply by lostanlen** [reply]
page 12 of this PDF
https://www.di.ens.fr/~mallat/papiers/ScatCPAM.pdf

Here's what happens when replacing scattering by Fourier modulus

```python
>>> np.linalg.norm(np.abs(np.fft.fft(x0)) - np.abs(np.fft.fft(xps))) / np.linalg.norm(x0)
0.004389970989466553

>>> plt.plot(np.fft.fftshift(np.abs(np.fft.fft(x0))-np.abs(np.fft.fft(xps))))
```

<img width="412" alt="Screenshot 2021-04-21 at 10 25 06" src="https://user-images.githubusercontent.com/3943142/115521970-f6d15580-a28b-11eb-9565-67ebd18cf4a7.png">

That's a pretty big artifact. I wonder where it comes from.

PS: setting `Q=1` will probably improve the invariance. (although we might not want to do that in practice)

**Reply by OverLordGoldDragon** [reply]
It's worse when FFT behaves just like one'd expect, `xps` being imaginary `x0`...

![image](https://user-images.githubusercontent.com/16495490/115629047-a5b87480-a312-11eb-94de-e760b1fbf0fa.png)

Immediate culprit: padding. And indeed:

```
l1l2_phase / l1l2_time: 0.5718327  # reflect
l1l2_phase / l1l2_time: 0.0025586  # zero
```

**Reply by lostanlen** [reply]
reflect padding considered harmful

**Comment by OverLordGoldDragon** [comment, marked as answer]
`reflect` bad

**Reply by OverLordGoldDragon** [reply]
I retried this with `max_pad_factor=None` (wasn't implemented at the time):

```
L1L2(time -shifted):    0.07115865385326169
L1L2(phase-shifted):    0.00035195798564448364
l1l2_phase / l1l2_time: 0.004946102358404171
```
`reflect` redeemed
